swarm/src/lib.rs

```diff
@@ -1121,6 +1121,12 @@ where
                     }
                 }
             }
+            ToSwarm::ListenOn { id, address } => {
+                self.transport.listen_on(id, address).ok()?;
```

Comment:
If this fails, the `NetworkBehaviour` will never learn about it. I think we need to capture this error and report it back to the `NetworkBehaviour` as well as issue a `ListenerError`.

The documentation on `ListenerError` talks about "non-fatal" errors but I think adding another variant is even more confusing.

Reply:
I was thinking about this back during the initial PR, but I assumed it would be alright since `ToSwarm::Dial` ignores the error. We could do this instead 

```suggestion
               if let Err(e) = self.transport.listen_on(id, address) {
                    self.behaviour.on_swarm_event(FromSwarm::ListenerError(
                        behaviour::ListenerError {
                            listener_id: id,
                            err: &e,
                        },
                    ));
                }
```

So it would notify the behaviour of the error instead of ignoring it.

Reply:
> I was thinking about this back during the initial PR, but I assumed it would be alright since `ToSwarm::Dial` ignores the error.

Are we? Probably also something that should be fixed then :)

> We could do this instead 
> 
> ```suggestion
>                if let Err(e) = self.transport.listen_on(id, address) {
>                     self.behaviour.on_swarm_event(FromSwarm::ListenerError(
>                         behaviour::ListenerError {
>                             listener_id: id,
>                             err: &e,
>                         },
>                     ));
>                 }
> ```
> 
> So it would notify the behaviour of the error instead of ignoring it. 

Yep, that looks good to me!

Reply:
> > I was thinking about this back during the initial PR, but I assumed it would be alright since `ToSwarm::Dial` ignores the error.
> 
> Are we? Probably also something that should be fixed then :)

Looking at `Swarm::dial`, it does send the error to the behaviour internally before returning the error itself so it doesnt look to be necessary

Reply:
That error will go away once we have https://github.com/multiformats/rust-multiaddr/issues/73.